pkg/apis/sessions/session_state.go

```diff
@@ -59,78 +65,119 @@ func (s *SessionState) String() string {
 	return o + "}"
 }
 
-// EncodeSessionState returns string representation of the current session
-func (s *SessionState) EncodeSessionState(c encryption.Cipher) (string, error) {
-	var ss SessionState
-	if c == nil {
-		// Store only Email and User when cipher is unavailable
-		ss.Email = s.Email
-		ss.User = s.User
-		ss.PreferredUsername = s.PreferredUsername
-	} else {
-		ss = *s
-		for _, s := range []*string{
-			&ss.Email,
-			&ss.User,
-			&ss.PreferredUsername,
-			&ss.AccessToken,
-			&ss.IDToken,
-			&ss.RefreshToken,
-		} {
-			err := into(s, c.Encrypt)
-			if err != nil {
-				return "", err
-			}
+// EncodeSessionState returns an encrypted, lz4 compressed, MessagePack encoded session
+func (s *SessionState) EncodeSessionState(c encryption.Cipher, compress bool) ([]byte, error) {
+	// Marshal to MessagePack
+	packed, err := msgpack.Marshal(s)
+	if err != nil {
+		return nil, err
+	}
+
+	if !compress {
+		// Encrypt the msgpack encoded
+		return c.Encrypt(packed)
+	}
+
+	// Compress the packed encoding
+	// The Compress:Decompress ratio is 1:Many. LZ4 gives fastest decompress speeds
+	buf := new(bytes.Buffer)
+	zw := lz4.NewWriter(nil)
+	zw.Header = lz4.Header{
+		BlockMaxSize:     65536,
+		CompressionLevel: 0,
+	}
+	zw.Reset(buf)
+
+	reader := bytes.NewReader(packed)
+	_, err = io.Copy(zw, reader)
+	if err != nil {
+		return nil, err
+	}
+	err = zw.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	compressed, err := ioutil.ReadAll(buf)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Might be nice to extract this to a function called something like `compress`? (or alternate because that would clash with the function parameter) Would make the flow of the function nicer I think
This function could look like
```
// Marshal to MessagePack
packed, err := msgpack.Marshal(s)
if err != nil {
	return nil, err
}

if compress {
	packed, err = compress(packed)
	if err != nil {
		return nil, err
	}
}

return c.Encrypt(packed)
```